### ISSUE TITLE
Node tree state is inconsistent during `NOTIFICATION_CHILD_ORDER_CHANGED`

### ISSUE THREAD

**Issue by RedworkDE** [godotengine/godot#77870]
### Godot version

543750a1b3f5696f9ba8e91cb49dc7db05d2ae62 

### Issue description

Looking into the cause of the `Children name does not match parent name in hashtable, this is a bug.` from #77845 and making notes in this issue before I forget what I wanted to report (again).

https://github.com/godotengine/godot/blob/543750a1b3f5696f9ba8e91cb49dc7db05d2ae62/scene/main/node.cpp#L1432-L1440

During `NOTIFICATION_CHILD_ORDER_CHANGED` the child node had been removed from the parents list of children, the child still has the parent set, making it possible to attempt to remove it from its parent again triggering the error.

### Steps to reproduce

Add this script to any node and run

```gdscript
extends Node

var node : Node

func _ready():
	var n = Node.new()
	add_child(n)
	node = n
	remove_child(n)

func _notification(what):
	if what == NOTIFICATION_CHILD_ORDER_CHANGED && node && node.get_parent() == self:
		remove_child(node)
```

### Minimal reproduction project

[Test_77870.zip](https://github.com/godotengine/godot/files/11651877/Test_77870.zip)

**Comment by Sauermann** [comment]
It is likely, that merging #77867 will hide this issue, so that a more elaborate setup will be required to trigger it.

**Comment by RedworkDE** [comment]
So while the current error was added in #75627 this issue is actually really old, and exists in some form since 2018 when `NOTIFICATION_MOVED_IN_PARENT` was first added in #23564.

**Comment by Omgesus** [comment]
Just got the bug mentioned in the main post ("Children name does not match parent name in hashtable, this is a bug.") after updating from 4.0.3. Running through the terminal, it threw a ton of warnings (like the one at the beginning, but I only kept the one instead of the dozens for brevity's sake), gave the said error as well as "Condition "data.parent" is true" and a backtrace. 

Figure add this here in case it helps in some way. 

```
WARNING: Parent path './MobileGates001/StaticBody3D' for node 'CollisionShape3D' has vanished when instantiating: 'res://Subscenes/Mansion/IndRooms/CentralRooms/mobile_elevator_gate.tscn'.
     at: instantiate (scene/resources/packed_scene.cpp:125)
ERROR: Children name does not match parent name in hashtable, this is a bug.
   at: remove_child (scene/main/node.cpp:1426)
ERROR: Condition "data.parent" is true.
   at: ~Node (scene/main/node.cpp:3519)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.1.beta1.official (828ec2c5d005b6499c7c4c88beaf81767d05614b)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /usr/lib/libc.so.6(+0x39ab0) [0x7f60c3bf0ab0] (??:0)
[2] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a1] (??:0)
[3] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[4] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2cbdd60] (??:0)
[5] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a4] (??:0)
[6] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[7] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2cbdd60] (??:0)
[8] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a4] (??:0)
[9] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[10] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2cbdd60] (??:0)
[11] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a4] (??:0)
[12] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[13] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2cbdd60] (??:0)
[14] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a4] (??:0)
[15] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[16] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2cbdd60] (??:0)
[17] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a4] (??:0)
[18] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[19] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2cbdd60] (??:0)
[20] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47776a4] (??:0)
[21] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x12d6c7a] (??:0)
[22] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x1dce661] (??:0)
[23] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x47e4051] (??:0)
[24] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2fbc1ab] (??:0)
[25] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2fbcda5] (??:0)
[26] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2d03d50] (??:0)
[27] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2d40d0e] (??:0)
[28] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2d42c7a] (??:0)
[29] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x2d69b05] (??:0)
[30] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0xea42b0] (??:0)
[31] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x4547e13] (??:0)
[32] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0x454905e] (??:0)
[33] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0xea495f] (??:0)
[34] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0xe34473] (??:0)
[35] /usr/lib/libc.so.6(+0x23850) [0x7f60c3bda850] (??:0)
[36] /usr/lib/libc.so.6(__libc_start_main+0x8a) [0x7f60c3bda90a] (??:0)
[37] /home/drew/Documents/Godot_v4.1-beta1_linux.x86_64() [0xe6534e] (??:0)
-- END OF BACKTRACE --
```

**Comment by RedworkDE** [comment]
@Omgesus Can you open a new issue for that and include a project, there seem to be bigger issues than just this (this issue shouldn't result in crashes)

**Comment by Omgesus** [comment]
@RedworkDE I think I figured things out. I left the room while the project was reimporting (1100+ files takes a little bit, even on good hardware :p), and when I came back Godot had closed. My guess is that it crashed while importing and a few scenes were corrupted, and then it crashed trying to load some bad data. I did a git reset on my project and reimported everything, and it has been working so far.
